src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,179 @@
+            return; // Recommend nothing (GitToolRecommendation = NONE)
+        }
+        final Jenkins jenkins = Jenkins.get();
+        GitTool tool = GitUtils.resolveGitTool(gitImplementation, jenkins, null, TaskListener.NULL);
+        if (tool != null) {
+            gitTool = tool.getGitExe();
+        }
+    }
+
+    /**
+     * Determine and estimate the size of a .git cached directory
+     * @param source: Use a {@link AbstractGitSCMSource} to access a cached Jenkins directory, we do not lock it.
+     * @return useCache
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private boolean setSizeFromCache(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        String cacheEntry = source.getCacheEntry();
+        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry);
+        if (cacheDir != null) {
+            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+            GitClient client = git.getClient();
+            if (!client.hasGitRepo()) {
+                useCache = false;
+            } else {
+                useCache = true;
+                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
+                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+            }
+        } else {
+            useCache = false;
+        }
+        return useCache;
+    }
+
+    /**
+     * Recommend a git implementation on the basis of the given size of a repository
+     * @param sizeOfRepo: Size of a repository (in KiBs)
+     * @return a git implementation, "git" or "jgit"
+     */
+    private String determineSwitchOnSize(Long sizeOfRepo) {
+        if (sizeOfRepo != 0L) {
+            if (sizeOfRepo >= SIZE_TO_SWITCH) {
+                return "git";
+            } else {
+                return "jgit";
+            }
+        }
+        return "NONE";
+    }
+
+    /**
+     * Other plugins can estimate the size of repository using this extension point
+     * The size is assumed to be in KiBs
+     */
+    public static abstract class RepositorySizeAPI implements ExtensionPoint {
+
+        public abstract boolean acceptsRemote(String remote);
+
+        public abstract Long getSizeOfRepository(String remote);
+
+        public static ExtensionList<RepositorySizeAPI> all() {
+            Jenkins jenkins = Jenkins.getInstanceOrNull();
+            if (jenkins == null) {
+                return null;
+            }
+            return jenkins.getExtensionList(RepositorySizeAPI.class);
+        }
+    }
+
+    private boolean setSizeFromAPI(String repoUrl) {
+        List<RepositorySizeAPI> acceptedRepository = Objects.requireNonNull(RepositorySizeAPI.all())
+                .stream()
+                .filter(r -> r.acceptsRemote(repoUrl))
+                .collect(Collectors.toList());
+
+        if (acceptedRepository.size() == 1) {
```

Comment:
We don't know how the implementer plugins will accept the remote/url. If there are more than one that match the url, what should we do? Take the first? return a false? My suggestion (and it's only a suggestion and I'm open to discuss:
```suggestion
        if (acceptedRepository.size() > 0) {
```

Reply:
Valid point. Currently the code chooses to avoid the whole situation by returning `false`. What if, instead of asking for a size of type long, we ask for a non-primitive custom class which encapsulates details like 
- size type (in KiBs, in MiBs), 
- from which provider (for ex: Github or GitLab)

Reply:
I have coded something like this in a different branch, I can raise a separate PR for you and others to review.

Reply:
you will still have the same question to answer: what will you return if more than one implementer extensions are marked as valid? The first one or null? Or do you will try one by one until one retrieves a valid object? those are the design question you have to try to answer

Reply:
After searching for an answer for sometime, I've come up with certain ideas: 
- **Accept a singleton instance**: Assume that it is not possible to have two extension instances of the same type, i.e, if the URL is from Github provider, assume that only Github Branch Source Plugin is going to implement it. It doesn't make sense to me that any other plugin would want to implement it. If we find out that more than one does, we **reject all instances**. 

Advantage: The extension point contract is simple at this moment, from git plugin's point of view, we can't assume to know what classes will implement it, so we ask them a simple question: Do you want to implement it? Yes or no. 
If we add another rule on the contract, a method which returns the type of provider implementing the extension point, it adds a little bit of complexity.

Disadvantage: We miss out on the correct instance which provides the size for our estimation.

- **Add another method on the contract**: Ask for the provider's name, say a method like `public abstract String returnProviderName()` which returns strings like "Github" or "Gitlab" which we can then use to filter out more than one instance.

Reply:
I'll try to answer bit by bit, but starting from the end:

> Add another method on the contract: Ask for the provider's name, say a method like public abstract String returnProviderName() which returns strings like "Github" or "Gitlab" which we can then use to filter out more than one instance.

I don't think that's a good idea because you would need a mechanism to distinguish the "provider" in the git-plugin itself and as Mark stated very well, it's better to leave that responsibility to the implementer. See what I mean:
- Let's suppose we add the `distinguishWhenUnclear` method.
- If implementer plugins have a really good `accept` method, so your URL is only accepted by 1 plugin, then your `distinguish` method is not actually needed.
- If implementer plugins have vague `accept` method, so more than one is accepting the URL, you have to distinguish between them. Then why do you need them to implement the `accept` method?

If we see the provider name is useful we can add it, but not for filtering purposes.

> Accept a singleton instance: Assume that it is not possible to have two extension instances of the same type, i.e, if the URL is from Github provider, assume that only Github Branch Source Plugin is going to implement it. It doesn't make sense to me that any other plugin would want to implement it. If we find out that more than one does, we reject all instances.

That's a valid solution of course. We can return a singleton object or a list filter in the Estimator class. Both solutions are equally valid and will lead to the same end. What I wanted to do was to make you think is about what behaviour we want when more than one Extension objects are accepting and why. You've given me the answer with:
> If we find out that more than one does, we reject all instances.

for the `what`, and

> Assume that it is not possible to have two extension instances of the same type, i.e, if the URL is from Github provider, assume that only Github Branch Source Plugin is going to implement it. It doesn't make sense to me that any other plugin would want to implement it

for the why. Now my opinion/suggestion
_I agree with `if the URL is from Github provider, assume that only Github Branch Source Plugin is going to implement it. It doesn't make sense to me that any other plugin would want to implement it`, but just an error in a pattern can make your implementation return a false positive. We can assume that an Extension that has returned a false positive will return an error when it tries to get the size using an API that is not valid for it, so if we return the full list of Extensions, it's enough if we accept the first one that is able to give us a number and not an error. This way we can distinguish between implementations and the probability of getting a false number is actually low. In other words: it's a less aggressive approach_
But that's *my opinion/suggestion* and it does not mean that it makes sense or it's the best approach. I'd like your thoughts about that and also I'd like other mentors's opinions here, @MarkEWaite @justinharringa @omkar-dsd

Reply:
> I don't think that's a good idea because you would need a mechanism to distinguish the "provider" in the git-plugin itself and as Mark stated very well, it's better to leave that responsibility to the implementer. See what I mean:

Understood, it might only increase complexity within the git plugin which is not be necessary.

@fcojfernandez That seems like a neat approach to me, assume that if there is a false positive, there is a very low chance that numerical size returned by it would be non-zero. 
If I've understood you correctly, if we have multiple implementations in the extension list, we check the size and only accept the non-zero size instance. 

This leads me to ask another question to you, what if some plugin implements our method of `getSizeFromRepository()` and does so incorrectly? Do we expect an exception thrown from the that erroneous method? I wouldn't assume that because our method definition doesn't throw any exception.
Should we expect a `size = 0` if things go wrong from the implementers side?

Reply:
A report might be empty, so 0 might or not be a wrong value per se, but I don't think it's a use case we should care.
Apart from that, adding a safe check just in case any exception that happens is not properly handled by the implementer plugin is just free. It's possible that the implementer is not checking the final endpoint, for example. IMO, better safe than sorry, and if the Extension throws an exception or return 0, then you can discarded the value in both situation.

regarding my comment
> a safe check just in case any exception that happens is not properly handled by the implementer plugin

When you delegate the responsibility, it's always a good idea to have a check. Keep that always in mind because even when you want the exception to be thrown, you can catch it, process it and re-throw it